packages/protocol/chainmanager/manager.go

```diff
@@ -148,6 +217,105 @@ func (m *Manager) Commitments(id commitment.ID, amount int) (commitments []*Comm
+		forkingCommitment = chain.ForkingPoint
+
+		if commitment, _ = m.Commitment(forkingCommitment.Commitment().PrevID()); commitment == nil {
+			return nil, errors.Wrapf(ErrCommitmentUnknown, "unknown parent of solid commitment %s", forkingCommitment.Commitment().ID())
+		}
+	}
+
+	return forkingCommitment, nil
+}
+
+// ForkedEventByForkingPoint returns the forking event generated by a peer for the given forking point.
+func (m *Manager) ForkedEventByForkingPoint(commitmentID commitment.ID) (forkedEvent *ForkDetectedEvent, exists bool) {
+	if storage := m.commitmentsByForkingPoint.Get(commitmentID.Index(), false); storage != nil {
+		return storage.Get(commitmentID)
+	}
+
+	return
+}
+
+func (m *Manager) SwitchMainChain(head commitment.ID) error {
+	commitment, _ := m.Commitment(head)
+	if commitment == nil {
+		return errors.Wrapf(ErrCommitmentUnknown, "unknown commitment %s", head)
+	}
+
+	return m.switchMainChainToCommitment(commitment)
+}
+
+func (m *Manager) switchMainChainToCommitment(commitment *Commitment) error {
+	forkingPoint, err := m.forkingPointAgainstMainChain(commitment)
+	if err != nil {
+		return err
+	}
+
+	//  commitment is already part of the main chain
+	if forkingPoint == nil {
+		return nil
+	}
+
+	parentCommitment, _ := m.Commitment(forkingPoint.Commitment().PrevID())
+	if parentCommitment == nil {
+		return errors.Wrapf(ErrCommitmentUnknown, "unknown parent of solid commitment %s", forkingPoint.ID())
+	}
+
+	// Separate the main chain by remove it from the parent
+	oldMainCommitment := parentCommitment.mainChild()
+
+	// For each forking point coming out of the main chain we need to reorg the children
+	for fp := commitment.Chain().ForkingPoint; ; {
+		fpParent, _ := m.Commitment(fp.Commitment().PrevID())
+
+		mainChild := fpParent.mainChild()
+		newChildChain := NewChain(mainChild)
+
+		if err := fpParent.setMainChild(fp); err != nil {
+			return err
+		}
+
+		for childWalker := walker.New[*Commitment]().Push(mainChild); childWalker.HasNext(); {
+			childWalker.PushAll(m.propagateReplaceChainToMainChild(childWalker.Next(), newChildChain)...)
+		}
+
+		if fp == forkingPoint {
+			break
+		}
+
+		fp = fpParent.Chain().ForkingPoint
+	}
+
+	// Delete all commitments after newer than our new head if the chain was longer than the new one
+	snapshotChain := m.SnapshotCommitment.Chain()
+	snapshotChain.dropCommitmentsAfter(commitment.ID().Index())
+
+	// Separate the old main chain by removing it from the parent
+	parentCommitment.deleteChild(oldMainCommitment)
+
+	// Cleanup the old tree hanging from the old main chain from our cache
+	if children := oldMainCommitment.Children(); len(children) != 0 {
+		for childWalker := walker.New[*Commitment]().PushAll(children...); childWalker.HasNext(); {
+			childWalker.PushAll(m.deleteAllChildrenFromCache(childWalker.Next())...)
+		}
+	}
+
+	return nil
+}
+
+func (m *Manager) Evict(index epoch.Index) {
+	//TODO: do we need to evict more stuff?
+	m.commitmentsByForkingPoint.Evict(index)
```

Comment:
We don't need an eviction lock here?

What about `commitmentsByID`?

Reply:
No need for a lock, because the epochStorage and the Storage have locks already.
Regarding the `commitmentsByID` we need to discuss. I think the idea was to keep them all in memory, but maybe @karimodm and @hmoog can chime in